### PR TITLE
Add Adan optimizer.

### PR DESCRIPTION
**Pull request by carlosgmartin**
Picking up #410.

### CODE REVIEW

File: optax/_src/alias.py [hunk 1]
```diff
@@ -177,6 +177,82 @@ def adadelta(
...
+  """The ADAptive Nesterov momentum algorithm (Adan).
+
+  Adan first reformulates the vanilla Nesterov acceleration to develop a new
+  Nesterov momentum estimation (NME) method, which avoids the extra overhead of
+  computing gradient at the extrapolation point. Then Adan adopts NME to
+  estimate the gradient's first- and second-order moments in adaptive gradient
+  algorithms for convergence acceleration.
+
+  References:
+    Xie et al, 2022: https://arxiv.org/abs/2208.06677
```
Comment (fabianp) [inline optax/_src/alias.py:199]: please put the links in the same format as the one in the adam docstring

File: optax/_src/transform.py [hunk 2]
```diff
@@ -594,6 +594,91 @@ def update_fn(updates, state, params=None):
...
+    b1: float = 0.98,
+    b2: float = 0.92,
+    b3: float = 0.99,
+    eps: float = 1e-8,
+    eps_root: float = 0.0,
+) -> base.GradientTransformation:
+  """Rescale updates according to the Adan algorithm.
+
+  References:
+    [Xie et al, 2022](https://arxiv.org/abs/2208.06677)
```
Comment (fabianp) [inline optax/_src/transform.py:615]: same as before, wrong format for references.

please build the doc locally and make sure that all docstring display correctly

File: optax/_src/alias.py [hunk 3]
```diff
@@ -177,6 +177,82 @@ def adadelta(
...
+      in RMSProp), to avoid dividing by zero when rescaling.
+    weight_decay: strength of the weight decay regularization.
+    mask: a tree with same structure as (or a prefix of) the params PyTree,
+      or a Callable that returns such a pytree given the params/updates.
+      The leaves should be booleans, `True` for leaves/subtrees you want to
+      apply the weight decay to, and `False` for those you want to skip. Note
+      that the Adam gradient transformations are applied to all parameters.
+
+  Returns:
+    the corresponding `GradientTransformation`.
```
Comment (fabianp) [inline optax/_src/alias.py:219]: I think that ```:class:`optax.GradientTransformation` ``` should result in a nicer clickable link in the webpage. please build locally and check this.

File: optax/_src/alias.py [hunk 4]
```diff
@@ -177,6 +177,82 @@ def adadelta(
...
+
+  Adan first reformulates the vanilla Nesterov acceleration to develop a new
+  Nesterov momentum estimation (NME) method, which avoids the extra overhead of
+  computing gradient at the extrapolation point. Then Adan adopts NME to
+  estimate the gradient's first- and second-order moments in adaptive gradient
+  algorithms for convergence acceleration.
+
+  References:
+    Xie et al, 2022: https://arxiv.org/abs/2208.06677
+
```
Comment (fabianp) [inline optax/_src/alias.py:200]: please also add the mathematical description of the algorithm as its done in (say) the adam docstring

File: optax/_src/transform.py [hunk 5]
```diff
@@ -594,6 +594,91 @@ def update_fn(updates, state, params=None):
...
+    grad_tm1 = otu.tree_zeros_like(params)  # previous gradient
+    return ScaleByAdanState(
+        count=jnp.zeros((), jnp.int32),
+        mu=mu,
+        nu=nu,
+        delta=delta,
+        grad_tm1=grad_tm1,
+    )
+
+  def update_fn(updates, state, params=None):
```
Comment (fabianp) [inline optax/_src/transform.py:642]: could you add a couple of docstrings explaining what mathematical operations are done on each block? that would be helpful to map the mathematical description of the algorithm to the implementation here

**Comment by fabianp** [comment]
thanks for the contribution @carlosgmartin ! Left some minor documentation nitpicks in there

File: optax/_src/transform.py (hunk 5 above)
Comment (carlosgmartin) [reply optax/_src/transform.py:644]: @fabianp Do you mean putting comments inside `update_fn`, or extending the docstring of `scale_by_adan`?

Comment (carlosgmartin) [reply optax/_src/transform.py:643]: I've edited `update_fn` to make it more closely reflect Algorithm 1 of https://arxiv.org/pdf/2208.06677v4#page=6, and added a docstring to reflect this.

**Comment by carlosgmartin** [comment]
@fabianp Done.

**Comment by fabianp** [comment]
looks good to me. just one last thing. could you please paste here a screenshot (or two, if it doesn't fit) of the rendering of the adan optimizer's docstring in the webpage? thanks!

**Comment by carlosgmartin** [comment]
@fabianp <details>

image

</details>

**Pull request merged**